File: server/core/src/main/scala/org/infinispan/server/core/CacheIgnoreAware.scala
package org.infinispan.server.core

import java.util

/**
 * @author gustavonalle
 * @since 8.1
 */
trait CacheIgnoreAware {

   private val ignoredCaches = new util.HashSet[String]()

   def setIgnoredCaches(cacheNames: java.util.Set[String]) = {
      this.synchronized {
         ignoredCaches.clear()
         ignoredCaches.addAll(cacheNames)
      }
   }

   def unignore(cacheName: String) = this.synchronized {
      ignoredCaches.remove(cacheName)
   }

   def ignoreCache(cacheName: String) = this.synchronized {
      ignoredCaches.add(cacheName)
   }

   def isCacheIgnored(cacheName: String) = this.synchronized {
      ignoredCaches.contains(cacheName)
   }

}

File: server/core/src/main/java/org/infinispan/server/core/CacheIgnoreAware.java
package org.infinispan.server.core;

import java.util.Collection;
import java.util.concurrent.CompletionStage;

import org.infinispan.manager.EmbeddedCacheManager;

/**
 * Defines an interface to be used when a cache is to be ignored by a server implementation.  Any implementation should
 * be thread safe and allow for concurrent methods to be invoked.
 * @author gustavonalle
 * @author wburns
 * @since 9.0
 */
public interface CacheIgnoreAware {

   /**
    * No longer ignore the given cache if it was before
    * @param cacheName the cache to now not ignore
    */
   CompletionStage<Boolean> unignore(String cacheName);

   /**
    * Ignores a given cache if it wasn't before
    * @param cacheName the cache to ignore
    */
   CompletionStage<Void> ignoreCache(String cacheName);

   /**
    * Queries whether the cache is ignored
    * @param cacheName the cache to see if it is ignored
    * @return whether or not the cache is ignored
    */
   boolean isCacheIgnored(String cacheName);

   /**
    * @return Set of all cache names currently ignored for the cache manager
    */
   CompletionStage<Collection<String>> getIgnoredCaches(String cacheManager);

   void initialize(EmbeddedCacheManager cacheManager);
}
